Northstar.CustomServers/mod/scripts/vscripts/mp/_stats.nut

```diff
@@ -12,67 +14,1083 @@ global function PreScoreEventUpdateStats
 global function PostScoreEventUpdateStats
 global function Stats_OnPlayerDidDamage
 
+struct {
+	table< string, array<string> > refs
+	table< string, array< void functionref( entity, float, string ) > > callbacks
+
+	table< entity, table< string, int > > cachedIntStatChanges
+	table< table< string, float > > cachedFloatStatChanges
+
+	table< entity, float > playerKills
+	table< entity, float > playerKillsPvp
+	table< entity, float > playerDeaths
+	table< entity, float > playerDeathsPvp
+
+	bool isFirstStrike = true
+} file
+
 void function Stats_Init()
 {
+	AddCallback_OnPlayerKilled( OnPlayerOrNPCKilled )
+	AddCallback_OnNPCKilled( OnPlayerOrNPCKilled )
+	AddCallback_OnPlayerRespawned( OnPlayerRespawned )
+	AddCallback_OnClientConnected( OnClientConnected )
+	AddCallback_OnClientDisconnected( OnClientDisconnected )
+	AddCallback_GameStateEnter( eGameState.Epilogue, OnEpilogueStarted )
+
+	thread HandleDistanceAndTimeStats_Threaded()
+	thread SaveStatsPeriodically_Threaded()
+}
+
+void function AddStatCallback( string statCategory, string statAlias, string statSubAlias, void functionref( entity, float, string ) callback, string subRef )
+{	
+	if ( !IsValidStat( statCategory, statAlias, statSubAlias ) )
+		throw "INVALID STAT: " + statCategory + " : " + statAlias + " : " + statSubAlias
```

Comment:
`format` is more efficient and better to read imho
```suggestion
		throw format( "INVALID STAT: %s : %s : %s", statCategory, statAlias, statSubAlias )
```

Reply:
Sure, but like this shouldnt ever get hit anyway, and could probably be removed altogether, was more a debug thing